File: backend/routers/teacher.py
from datetime import date, datetime
from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session
from pydantic import BaseModel
from ..database import get_db
from ..deps import require_role
from ..models import Student, Attendance, User
from ..excel_export import build_report
from ..notifications import notify_parent

router = APIRouter(prefix="/api/teacher", tags=["teacher"])
teacher_only = require_role("Teacher", "Admin")


class ManualMark(BaseModel):
    student_id: int
    session_type: str    # Check-In / Check-Out
    status: str          # Present / Absent / Manual Override


@router.get("/roster")
def roster(class_section: str = None, db: Session = Depends(get_db),
           user: User = Depends(teacher_only)):
    q = db.query(Student)
    if class_section:
        q = q.filter(Student.class_section == class_section)
    today = date.today()
    result = []
    for s in q.all():
        recs = db.query(Attendance).filter(
            Attendance.student_id == s.id, Attendance.date == today).all()
        checkin = next((r.status for r in recs if r.session_type == "Check-In"), "Absent")
        checkout = next((r.status for r in recs if r.session_type == "Check-Out"), "Absent")
        result.append({"id": s.id, "name": s.name, "roll_number": s.roll_number,
                       "class_section": s.class_section,
                       "checkin": checkin, "checkout": checkout})
    return result


@router.post("/mark")
def mark(data: ManualMark, db: Session = Depends(get_db),
         user: User = Depends(teacher_only)):
    student = db.query(Student).filter(Student.id == data.student_id).first()
    if not student:
        raise HTTPException(404, "Student not found")
    today = date.today()
    existing = db.query(Attendance).filter(
        Attendance.student_id == data.student_id, Attendance.date == today,
        Attendance.session_type == data.session_type).first()
    now = datetime.now()
    if existing:
        existing.status = data.status
        existing.marked_by = f"Teacher_{user.full_name}"
        existing.timestamp = now
    else:
        existing = Attendance(student_id=data.student_id, date=today, timestamp=now,
                              session_type=data.session_type, status=data.status,
                              marked_by=f"Teacher_{user.full_name}")
        db.add(existing)
    db.commit()
    if data.status in ("Present", "Manual Override") and student.parent_id:
        notify_parent(db, student.parent_id, student.id, student.name,
                      data.session_type, now)
    return {"message": f"{student.name} marked {data.status} ({data.session_type})"}


@router.get("/logs")
def logs(db: Session = Depends(get_db), user: User = Depends(teacher_only)):
    recs = db.query(Attendance).order_by(Attendance.timestamp.desc()).limit(100).all()
    out = []
    for r in recs:
        out.append({"student": r.student.name if r.student else "?",
                    "roll": r.student.roll_number if r.student else "?",
                    "date": r.date.strftime("%d-%m-%Y"),
                    "time": r.timestamp.strftime("%I:%M %p"),
                    "session": r.session_type, "status": r.status,
                    "marked_by": r.marked_by})
    return out


@router.get("/export")
def export(db: Session = Depends(get_db), user: User = Depends(teacher_only)):
    buf = build_report(db)
    return StreamingResponse(
        buf,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": "attachment; filename=Class_Report.xlsx"})

File: frontend/js/teacher.js
const auth = guard("Teacher");
if(auth) document.getElementById("uname").textContent = auth.name;

function badge(status){
  if(status === "Present") return '<span class="badge badge-present">Present</span>';
  if(status === "Manual Override") return '<span class="badge badge-override">Override</span>';
  return '<span class="badge badge-absent">Absent</span>';
}

async function loadRoster(){
  const list = await api("/api/teacher/roster");
  document.getElementById("rosterRows").innerHTML = list.map(s=>`
    <tr><td>${s.roll_number}</td><td>${s.name}</td><td>${s.class_section}</td>
    <td>${badge(s.checkin)}</td><td>${badge(s.checkout)}</td>
    <td>
      <button class="btn btn-sm btn-green" onclick="mark(${s.id},'Check-In')">✓ In</button>
      <button class="btn btn-sm" onclick="mark(${s.id},'Check-Out')">✓ Out</button>
      <button class="btn btn-sm btn-red" onclick="absent(${s.id})">✗ Absent</button>
    </td></tr>`).join("");
}

async function mark(id, session){
  try {
    const r = await api("/api/teacher/mark", { method:"POST", body: JSON.stringify({
      student_id:id, session_type:session, status:"Manual Override" })});
    toast(r.message); loadRoster();
  } catch(e){ toast(e.message, "err"); }
}
async function absent(id){
  try {
    await api("/api/teacher/mark", { method:"POST", body: JSON.stringify({
      student_id:id, session_type:"Check-In", status:"Absent" })});
    toast("Marked Absent"); loadRoster();
  } catch(e){ toast(e.message, "err"); }
}

async function loadLogs(){
  const list = await api("/api/teacher/logs");
  document.getElementById("logRows").innerHTML = list.map(l=>`
    <tr><td>${l.date}</td><td>${l.time}</td><td>${l.roll}</td><td>${l.student}</td>
    <td>${l.session}</td><td>${badge(l.status)}</td><td>${l.marked_by}</td></tr>`).join("");
}

async function exportClass(){
  const a = getAuth();
  const res = await fetch("/api/teacher/export", { headers:{ "Authorization":"Bearer "+a.token }});
  const blob = await res.blob();
  const url = URL.createObjectURL(blob);
  const link = document.createElement("a"); link.href = url; link.download = "Class_Report.xlsx"; link.click();
  toast("Excel downloaded ✅");
}

loadRoster();